hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -826,26 +834,41 @@ update env ledger st ev = case (st, ev) of
     -- TODO: Is it really intuitive that we respond from the confirmed ledger if
     -- transactions are validated against the seen ledger?
     Effects [ClientEffect . ServerOutput.GetUTxOResponse headId $ getField @"utxo" $ getSnapshot confirmedSnapshot]
-  (Open openState, NetworkEvent _ _ (ReqDec{transaction})) ->
-    onOpenNetworkReqDec openState transaction
   (Open OpenState{headId, coordinatedHeadState, currentSlot}, ClientEvent Decommit{decommitTx}) -> do
-    -- TODO: Spec: require U̅ ◦ decTx /= ⊥
-    requireValidDecommitTx $ \utxoToDecommit ->
-      Effects
-        [ ClientEffect ServerOutput.DecommitRequested{headId, utxoToDecommit}
-        , NetworkEffect ReqDec{transaction = decommitTx}
-        ]
+    case mExistingDecommitTx of
+      Just existingDecommitTx ->
+        Effects
+          [ ClientEffect
+              ServerOutput.DecommitInvalid
+                { headId
+                , decommitInvalidReason = ServerOutput.DecommitAlreadyInFlight{decommitTx = existingDecommitTx}
+                }
+          ]
+      Nothing ->
+        requireValidDecommitTx headId ledger currentSlot coordinatedHeadState decommitTx $
+          Effects
+            [ NetworkEffect ReqDec{transaction = decommitTx, decommitRequester = party}
+            ]
    where
-    -- TODO: Spec: require U̅ ◦ decTx /= ⊥
-    requireValidDecommitTx cont =
-      case applyTransactions ledger currentSlot confirmedUTxO [decommitTx] of
-        Left (_, err) ->
-          Error $ RequireFailed $ DecommitTxInvalid{decommitTx, error = err}
-        Right _ -> cont $ utxoFromTx decommitTx
-
-    confirmedUTxO = (getSnapshot confirmedSnapshot).utxo
-
-    CoordinatedHeadState{confirmedSnapshot} = coordinatedHeadState
+    Environment{party} = env
+    CoordinatedHeadState{decommitTx = mExistingDecommitTx} = coordinatedHeadState
+  (Open openState@OpenState{coordinatedHeadState}, NetworkEvent _ _ (ReqDec{transaction})) ->
+    case mExistingDecommitTx of
+      Just existingDecommitTx ->
+        Error $ RequireFailed $ DecommitTxInFlight{decommitTx = existingDecommitTx}
+      Nothing -> onOpenNetworkReqDec env openState transaction
```

Comment:
Could: keep the whole logic in the handler function or none (to stay consistent with the other handlers)

I was caught off guard when reviewing this that this check is here and not in `onOpenNetworkReqDec`